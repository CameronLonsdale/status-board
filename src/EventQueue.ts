/**
 * Event Queue
 *
 * @param io
 * @constructor
 */
export default class EventQueue {
  private io: any;
  private latestEvents: any;

  constructor(io: any) {
    this.io = io;
    this.latestEvents = {};

    io.on('connection', (socket: any) => {
      socket.on('resend', (data: any) => {
        if (this.latestEvents[data]) {
          socket.emit(data, this.latestEvents[data]);
        }
      });

      // broadcast logs
      socket.on('log', (data: any) => {
        socket.broadcast.emit('client', data);
      });
    });
  }

  /**
   * Send widget data to clients
   * @param id
   * @param data
   */
  public send(id: any, data: any) {
    this.latestEvents[id] = data;
    this.io.emit(id, data); // emit to widget
    this.io.emit('client', { data, widgetId: id }); // emit to logger
  }
}
