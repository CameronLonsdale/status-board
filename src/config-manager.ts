import * as bug from 'debug';
import * as fs from 'fs';
import * as path from 'path';
import * as extend from 'xtend';
import { logger } from './logger';

const debug = bug('config-manager');

function readConfigIfExists(fileName: string) {
  let configFileName = fileName;

  if (!path.extname(configFileName)) {
    configFileName = configFileName + '.js';
  }

  if (fs.existsSync(configFileName)) {
    return require(configFileName);
  }

  return {};
}

function readEnv(configFileName: string) {
  const key = 'ATLASBOARD_CONFIG_' + configFileName;

  debug('ENV key', key);

  const environmentKey = process.env[key];

  if (environmentKey) {

    debug('ENV configuration found for', key);

    try {
      const configValue = JSON.parse(environmentKey);

      if (typeof configValue === 'object') {
        return JSON.parse(environmentKey);
      }

      logger.error(`
        ENV configuration key ${key} could not be serialized into an object: ${environmentKey}
      `);

    } catch (error) {
      throw Error(`
        ENV configuration key ${key} contains invalid JSON: ${environmentKey}
        Error: ${error}
      `);
    }
  }
}

export default function (configFileName: any) {
  const localConfigFilePath = path.join(process.cwd(), 'config', configFileName);
  const atlasboardConfigFilePath = path.join(__dirname, '../config/', configFileName);

  return extend(
    readConfigIfExists(atlasboardConfigFilePath),
    readConfigIfExists(localConfigFilePath),
    readEnv(configFileName),
  );
}
