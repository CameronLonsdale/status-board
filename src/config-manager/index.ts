import * as path from 'path';
import * as extend from 'xtend';
import { readConfigIfExists } from './read-config-if-exists';
import { readEnv } from './read-env';

export function configManager(configFileName: any) {
  const localConfigFilePath = path.join(process.cwd(), 'config', configFileName);
  const atlasboardConfigFilePath = path.join(__dirname, '../../config/', configFileName);

  return extend(
    readConfigIfExists(atlasboardConfigFilePath),
    readConfigIfExists(localConfigFilePath),
    readEnv(configFileName),
  );
}
