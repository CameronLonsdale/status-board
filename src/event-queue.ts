/**
 * Event Queue
 *
 * @param io
 * @constructor
 */
export default function eventQueue(io: any) {
  this.io = io;
  this.latestEvents = {};

  io.on('connection', (socket: any) => {
    socket.on('resend', (data: any) => {
      if (this.latestEvents[data]) {
        socket.emit(data, this.latestEvents[data]);
      }
    });

    // broadcast logs
    socket.on('log', (data: any) => {
      socket.broadcast.emit('client', data);
    });
  });

  return {
    /**
     * Send widget data to clients
     * @param id
     * @param data
     */
    send: (id: any, data: any) => {
      this.latestEvents[id] = data;
      this.io.emit(id, data); // emit to widget
      this.io.emit('client', { data, widgetId: id }); // emit to logger
    },
  };
}
