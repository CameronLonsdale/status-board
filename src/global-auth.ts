import * as fs from 'fs';
import * as traverse from 'traverse';

import logger from './logger';

const ENV_VAR_REGEX = /\$\{([^}]+)\}/;

export function globalAuth(file: any) {
  let auth = {};

  try {
    auth = JSON.parse(fs.readFileSync(file).toString());
  } catch (e) {
    if (e.code === 'ENOENT') {
      logger().warn(`Authentication file not found in ${file}.`);
      logger().warn('You may want to create your own.');
      logger().warn('You can also define the place where the credential file will be located by ' +
        'editing the auth file configuration property \'authenticationFilePath\'');
    } else {
      logger().error('Error reading ' + file + '. It may contain invalid json format');
    }
    return auth;
  }

  try {
    traverse(auth).forEach(function (val) {
      if ('string' === typeof val) {
        const match = ENV_VAR_REGEX.exec(val);
        let modified;
        let newVal;
        if (match !== null) {
          const envName = match[1];
          let envVal = process.env[envName];
          if (envVal === undefined) {
            // tslint:disable-next-line max-line-length
            logger().warn(`Authentication file referenced var \${${envName}}, which was not present in environment.`);
            envVal = '';
          }

          newVal = val.substring(
            0,
            match.index,
          ) + envVal + val.substring(
            match.index + match[0].length,
          );
          modified = true;
        }

        if (modified) {
          this.update(newVal);
        } else {
          this.update(val);
        }
      }
    });
  } catch (error) {
    throw Error(`
      Error parsing the auth file ${file} with env variables.
      Error: ${error}
    `);
  }

  return auth;
}
