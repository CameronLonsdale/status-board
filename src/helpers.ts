import * as fs from 'fs';

export function readJSONFile(dashboardPath: any, cb: any) {
  fs.readFile(dashboardPath, (error: any, data: any) => {
    if (error) {
      return cb(error);
    }

    try {
      cb(null, JSON.parse(data));
    } catch (error) {
      cb(error);
    }
  });
}

export function noop() {
  // No operation performed.
}
