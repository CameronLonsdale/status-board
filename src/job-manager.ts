import * as fs from 'fs';
import * as path from 'path';
import * as extend from 'xtend';
import { get, resolveCandidates } from './item-manager';

/**
 * Return a particular dashboard object
 *
 * @param  {string} dashboardFilePath dashboard path
 * @return {object} dashboard object
 */
function readDashboard(dashboardFilePath: any) {
  const dashboardConfig = JSON.parse(fs.readFileSync(dashboardFilePath));

  if (!dashboardConfig.layout) {
    throw('No layout field found in ' + dashboardFilePath);
  }

  if (!dashboardConfig.layout.widgets) {
    throw('No widgets field found in ' + dashboardFilePath);
  }
  return dashboardConfig;
}

/**
 * Returns true if dashboard matches a particular regex filter
 *
 * @param  {string} dashboardFullPath dashboard full path
 * @param  {string} filter regex
 * @return {boolean}
 */
function matchDashboardFilter(dashboardFullPath: any, filter: any) {
  const dashboardName = path.basename(dashboardFullPath);
  return dashboardName.match(filter);
}

/**
 * Returns true if job matches a particular regex filter
 *
 * @param  {string} jobName job name
 * @param  {string} filter regex
 * @return {boolean}
 */
function matchJobFilter(jobName: any, filter: any) {
  return jobName.match(filter);
}

/**
 * Process all jobs from a dashboard
 *
 * @param  {array} allJobs all available jobs
 * @param  {string} dashboardName dashboard name
 * @param  {object} dashboardConfig dashboard config
 * @param  {object} filters filters, if any
 * @return {array} related jobs
 */
function processDashboard(allJobs: any, dashboardName: any, dashboardConfig: any, filters: any) {
  const jobs = [];
  for (let i = 0, l = dashboardConfig.layout.widgets.length; i < l; i++) {
    const jobItem = dashboardConfig.layout.widgets[i];
    if (jobItem.job) { // widgets can run without a job, displaying just static html.
      if (filters.jobFilter) {
        if (!matchJobFilter(jobItem.job, filters.jobFilter)) {
          continue;
        }
      }

      const candidateJobs = resolveCandidates(allJobs, jobItem.job, 'jobs', '.js');
      if (!candidateJobs.length) {
        throw '  ERROR RESOLVING JOB ' +
        '\n   No job file found for "' + jobItem.job + '" in ' + dashboardName +
        '\n   Have you pulled all the packages dependencies? (They are git submodules.)' +
        '\n\n   $ git submodule init' +
        '\n   $ git submodule update\n';
      }

      const job = {
        configKey: jobItem.config,
        dashboard_name: path.basename(dashboardName, '.json'),
        job_name: jobItem.job,
        widget_item: jobItem,
      };

      const jobRequire = require(candidateJobs[0]);
      if (typeof jobRequire === 'function') {
        job.onRun = jobRequire;
      } else {
        job.onRun = jobRequire.onRun || function () {
        };
        job.onInit = jobRequire.onInit || function () {
        };
      }

      jobs.push(job);
    }
  }
  return jobs;
}

export default {
  /**
   * Return the jobs for all available dashboards in all the packages
   *
   * @param  {object}   options  options object
   * @param  {Function} callback
   */
  getJobs(options: any, callback: any) {

    const packagesPath = options.packagesPath;
    const filters = options.filters || {};

    const configPath = path.join(options.configPath, '/dashboard_common.json');
    let generalDashboardConfig = {};

    let jobs = [];

    // ----------------------------------------------
    // general config is optional, but if it exists it needs to be a valid file
    // ----------------------------------------------
    if (fs.existsSync(configPath)) {
      try {
        generalDashboardConfig = JSON.parse(fs.readFileSync(configPath)).config;
        if (!generalDashboardConfig) {
          throw 'invalid format. config property not found';
        }
      } catch (e) {
        return callback('ERROR reading general config file...' + configPath);
      }
    }

    // ----------------------------------------------
    // get all dashboards from all packages folder
    // ----------------------------------------------
    get(packagesPath, 'dashboards', '.json', (err: any, dashboardConfigFiles: any) => {
      if (err) {
        return callback(err);
      }

      // ----------------------------------------------
      // get all jobs from those packages
      // ----------------------------------------------
      get(packagesPath, 'jobs', '.js', (err: any, allJobs: any) => {
        if (err) {
          return callback(err);
        }

        for (let d = 0, dl = dashboardConfigFiles.length; d < dl; d++) {
          const dashboardName = dashboardConfigFiles[d];

          if (filters.dashboardFilter) {
            if (!matchDashboardFilter(dashboardName, filters.dashboardFilter)) {
              continue;
            }
          }

          let dashboardConfig;
          let dashboardJobs;
          try {
            dashboardConfig = readDashboard(dashboardName);
            dashboardJobs = processDashboard(allJobs, dashboardName, dashboardConfig, filters);
          } catch (e) {
            return callback(e);
          }

          // add config to job, extending for the same config key in general config, if any
          dashboardJobs = dashboardJobs.map((job) => {
            // Multiple configurations:
            //  local overrides global
            //  config n+1 overrides config n
            if (Array.isArray(job.configKey)) {
              const configs = job.configKey.map((key) => {
                return extend(generalDashboardConfig[key], dashboardConfig.config[key]);
              });
              job.config = extend.apply(null, configs);
            } else { // single configuration
              job.config = extend(generalDashboardConfig[job.configKey], dashboardConfig.config[job.configKey]);
            }

            return job;
          });

          jobs = jobs.concat(dashboardJobs);
        }

        callback(null, jobs);
      });
    });
  },
};
