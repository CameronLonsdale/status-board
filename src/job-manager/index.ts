export { getJobs } from './get-jobs';
