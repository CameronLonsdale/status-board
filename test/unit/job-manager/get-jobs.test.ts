import { getJobs } from '../../../src/job-manager';

describe('Job Manager: Get Jobs', () => {
  test('Should match snapshot', () => {
    expect(getJobs).toMatchSnapshot();
  });
});
