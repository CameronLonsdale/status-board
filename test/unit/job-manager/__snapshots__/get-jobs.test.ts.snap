// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Job Manager: Get Jobs Should match snapshot 1`] = `[Function]`;
